File: employee-management-backend/services/authService.js
const repository =
    require("../repositories/authRepository");

const passwordUtil =
    require("../utils/passwordUtil");

const jwtUtil =
    require("../utils/jwtUtil");

const auditService =
    require("./auditService");

exports.login =
    async (
        username,
        password
    ) => {

        const users =
            await repository.findByUsername(
                username
            );

        if (
            !users ||
            users.length === 0
        ) {

            throw new Error(
                "Invalid Username"
            );

        }

        const user =
            users[0];

        const valid =
            await passwordUtil.comparePassword(
                password,
                user.PASSWORD
            );

        if (!valid) {

            throw new Error(
                "Invalid Password"
            );

        }

        // Audit Login
        await auditService.log(

            {
                userId:
                    user.USER_ID,

                username:
                    user.USERNAME
            },

            "LOGIN",

            "APP_USERS",

            user.USER_ID,

            "User Logged In"

        );

        return {

            token:
                jwtUtil.generateToken(
                    user
                ),

            username:
                user.USERNAME,

            role:
                user.ROLE

        };

    };

exports.register =
    async (user) => {

        const existing =
            await repository.findByUsername(
                user.username
            );

        if (
            existing &&
            existing.length > 0
        ) {

            throw new Error(
                "Username already exists"
            );

        }

        const hash =
            await passwordUtil.hashPassword(
                user.password
            );

        await repository.createUser({

            username:
                user.username,

            password:
                hash,

            fullName:
                user.fullName,

            email:
                user.email,

            role:
                user.role ||
                "EMPLOYEE"

        });

        // Audit Register
        await auditService.log(

            null,

            "REGISTER",

            "APP_USERS",

            user.username,

            "New User Registered"

        );

        return true;

    };

File: employee-management-springboot/src/main/java/com/ems/service/AuthService.java
package com.ems.service;

import com.ems.dto.LoginRequest;
import com.ems.dto.LoginResponse;
import com.ems.dto.RegisterRequest;
import com.ems.exception.AppException;
import com.ems.model.AppUser;
import com.ems.repository.AuthRepository;
import com.ems.security.JwtUtil;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
@Slf4j
public class AuthService {

    private final AuthRepository  authRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtUtil         jwtUtil;

    public LoginResponse login(LoginRequest req) {
        log.debug("AuthService.login - username: {}", req.getUsername());
        
        AppUser user = authRepository.findByUsername(req.getUsername())
                .orElseThrow(() -> {
                    log.warn("User not found: {}", req.getUsername());
                    return new AppException("Invalid username or password", HttpStatus.UNAUTHORIZED);
                });

        log.debug("User found: {} with status: {}", user.getUsername(), user.getStatus());

        if (!"ACTIVE".equalsIgnoreCase(user.getStatus())) {
            log.warn("Account inactive for user: {}", user.getUsername());
            throw new AppException("Account is inactive", HttpStatus.UNAUTHORIZED);
        }

        log.debug("Comparing passwords...");
        if (!passwordEncoder.matches(req.getPassword(), user.getPassword())) {
            log.warn("Invalid password for user: {}", user.getUsername());
            throw new AppException("Invalid username or password", HttpStatus.UNAUTHORIZED);
        }

        log.debug("Generating token for user: {} with role: {}", user.getUsername(), user.getRole());
        String token = jwtUtil.generateToken(user.getUserId(), user.getUsername(), user.getRole());
        return new LoginResponse(token, user.getUsername(), user.getRole());
    }

    public AppUser register(RegisterRequest req) {
        if (authRepository.existsByUsername(req.getUsername())) {
            throw new AppException("Username already exists", HttpStatus.CONFLICT);
        }

        String hashedPassword = passwordEncoder.encode(req.getPassword());

        AppUser user = new AppUser();
        user.setUsername(req.getUsername());
        user.setPassword(hashedPassword);
        user.setFullName(req.getFullName());
        user.setEmail(req.getEmail());
        user.setRole(req.getRole() != null ? req.getRole() : "EMPLOYEE");

        return authRepository.createUser(user);
    }

    public AppUser getMe(Long userId) {
        return authRepository.findById(userId)
                .orElseThrow(() ->
                        new AppException("User not found", HttpStatus.NOT_FOUND));
    }
}
